dunai/src/Data/MonadicStreamFunction/Instances/Profunctor.hs

```diff
@@ -0,0 +1,45 @@
+{-# LANGUAGE CPP                 #-}
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+-- | Instances of 'Profunctor' and 'Traversing' for Monadic Stream Functions ('MSF').
+--
+--   Import this module to include the (orphan) instances.
+module Data.MonadicStreamFunction.Instances.Profunctor where
+
+-- base
+import Control.Arrow
+
+#if !MIN_VERSION_base(4,8,0)
+import Data.Traversable (traverse)
+import Control.Applicative (Applicative)
+#endif
+
+-- profunctors
+import Data.Profunctor
+import Data.Profunctor.Traversing
+
+-- dunai
+import Data.MonadicStreamFunction.Core
+import Data.MonadicStreamFunction.InternalCore
+import Data.MonadicStreamFunction.Instances.ArrowChoice ()
+
+-- transformers
+import Control.Monad.Trans.State.Strict
+
+instance Monad m => Choice (MSF m) where
+  left' = left
+
+instance Monad m => Strong (MSF m) where
+  first' = first
+
+instance Monad m => Profunctor (MSF m) where
+  dimap l r p = arr l >>> p >>> arr r
+
+instance (Functor m, Applicative m, Monad m) => Traversing (MSF m) where
+  traverse' (MSF msf) =
+    let unMSF' x = StateT (\(MSF m) -> m x)
```

Comment:
You could also write `unMSF' = StateT . flip unMSF`. I'd personally prefer that style, and then inline it in line 44, but others might not.